pkg/csi_driver/controller.go

```diff
@@ -101,15 +106,66 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
 	} else {
+
+		// Get the reserved-IPV4-cidr from the request. If value is not provided, reservedIPV4Range is defaulted to ""
+		reservedIPV4CIDR, ok := req.GetParameters()[paramReservedIPV4CIDR]
+		if ok {
+
+			// Prevents lock acquisition in case the CIDR is invalid
+			validCIDR := s.config.ipAllocator.ValidateCIDR(reservedIPV4CIDR)
+			if !validCIDR {
+				return nil, err
+			}
+
+			reservedIPRange, err := s.reserveIPV4Range(ctx, newFiler, reservedIPV4CIDR)
+			if err != nil {
+				return nil, err
+			}
+
+			// Adding the reserved IP range to the instance object
+			newFiler.Network.ReservedIpRange = reservedIPRange
+		}
+
 		// Create the instance
 		filer, err = s.config.fileService.CreateInstance(ctx, newFiler)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 	}
+
+	//Free up the reserved IPV4 Range from the ipAllocator as we would be recieving this value in the listInstances call as the createInstance operation was completed
+	s.config.ipAllocator.Unreserve(filer.Network.ReservedIpRange)
 	return &csi.CreateVolumeResponse{Volume: fileInstanceToCSIVolume(filer, modeInstance)}, nil
 }
 
+// getUnreservedIPBlock returns the available IP in the cidr. It is synchronized in order to avoid data race while getting unreserved IP range in a CIDR
+func (s *controllerServer) reserveIPV4Range(ctx context.Context, filer *file.ServiceInstance, cidr string) (string, error) {
+	instances, err := s.config.fileService.ListInstances(ctx, filer)
+	if err != nil {
+		return "", status.Error(codes.Aborted, err.Error())
+	}
+
+	// Initialize an empty reserved list. It will be populated with all the reservedIPRanges obtained from the active instances
+	reservedIPRanges := make(map[string]bool)
+	for _, instance := range instances {
+		reservedIPRanges[instance.Network.ReservedIpRange] = true
+	}
+
+	// Getting unreserved IP in a CIDR needs to be synchronized
+	s.config.mutex.Lock()
+	defer s.config.mutex.Unlock()
+
+	// Pass the reservedIPRanges obtained from ListInstances call. We have the IPRanges reserved at the service end in the IPAllocator.reserved list. The GetUnreservedIPBlock would merge these two lists to form a reserved list
+	unreservedIPBlock, err := s.config.ipAllocator.GetUnreservedIPBlock(cidr, reservedIPRanges)
+	if err != nil {
+		return "", err
+	}
+
+	// Reserve the IPRange obtained from the GetUnreservedIPBlock call
+	s.config.ipAllocator.Reserve(unreservedIPBlock)
```

Comment:
Wanted to decouple the Get and Reserve Operations. Also, adding a mutating operation in Get conflicts with the naming.

Reply:
From the user's perspective, is there a purpose for getting an IP block without reserving it?  If not, then I think it can be condensed into two interfaces.  Internally, it could remain as two different functions.